Support generic associated consts

michaelwu opened the pull request:
This provides limited support for using associated consts on type parameters. It generally works on things that can be figured out at trans time. This doesn't work for array lengths or match arms. I have another patch to make it work in const expressions.

CC @eddyb @nikomatsakis

rust-highfive commented:
Thanks for the pull request, and welcome! The Rust team is excited to review your changes, and you should hear from @pnkfelix (or someone else) soon.

If any changes to this PR are deemed necessary, please add them as extra commits. This ensures that the reviewer can see what has changed since they last reviewed the code. Due to the way GitHub handles out-of-date commits, this should also make it reasonably obvious what issues have or haven't been addressed. Large or tricky changes may require several passes of review and changes.

Please see [the contribution instructions](https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md) for more information.

michaelwu commented:
I think @nikomatsakis wanted to review this - didn't realize rust-highfive automatically assigned a reviewer.

oli-obk commented:
the corresponding RFC has not yet been accepted: https://github.com/rust-lang/rfcs/pull/1062

michaelwu commented:
FWIW, this does a different thing than what the RFC proposes. That RFC attempts to make certain tricky situations work like array lengths and match patterns work. This doesn't do any of that - it only supports the parts that are simple to handle.

nikomatsakis commented:
r? @nikomatsakis

nikomatsakis commented:
stole review since @michaelwu and I already looked over this at work week

nikomatsakis commented on src/librustc/middle/const_eval.rs:
```diff
@@ -1265,12 +1273,8 @@ fn resolve_trait_associated_const<'a, 'tcx: 'a>(tcx: &'a ty::ctxt<'tcx>,
-                                        constant trait item reference.",
-                                       e))
+        Err(_) => {
+            return None
```
@michaelwu why did you change this to `None`?

michaelwu replied:
IIRC this was to avoid hitting an error in early const eval. `fn select()` returns `Err(_)` if there's a substitution that can't be found. Maybe it should return `Ok(None)`?

nikomatsakis replied:
Hmm. Yes, I suspect calling `bug` was wrong. I don't have a great grasp on where an error on the user's part will fall out here, but I guess I would expect to see it come out during typeck. I should check if there are existing test cases.

nikomatsakis commented on src/test/run-pass/associated-const-type-parameters.rs:
```diff
@@ -0,0 +1,41 @@
+}
+
+fn main() {
+    println!("{}", Abc::X);
```
can you rewrite this test to use asserts?

nikomatsakis commented on src/test/run-pass/associated-const-type-parameters.rs:
```diff
@@ -0,0 +1,41 @@
+    println!("{}", Abc::get_x());
+    println!("{}", Def::get_x());
+    println!("{}", add::<Abc, Def>());
+    assert_eq!(108, add::<Abc, Def>());
```
like this :)

nikomatsakis replied:
also, can you change this from `add` to `sub`, so that we know that the order is correct? that way, we can also test the result of `sub::<Def, Abc>()`, for example.

nikomatsakis commented on src/librustc/middle/const_eval.rs:
```diff
@@ -90,7 +91,8 @@ fn lookup_variant_by_id<'a>(tcx: &'a ty::ctxt,
                                         def_id: DefId,
-                                        maybe_ref_id: Option<ast::NodeId>)
+                                        maybe_ref_id: Option<ast::NodeId>,
+                                        param_substs: Option<&'tcx subst::Substs<'tcx>>)
```
This merits a comment, I think, as to the purpose of each of these parameters. Something like:

- `def_id` is the id of the constant
- `maybe_ref_id` is the id of the expression referencing the constant, if available
- `param_substs` is the monomorphization substitution for the expression, if available

The latter two parameters are used to extract the substitutions for the constant, in the case of associated constants.

nikomatsakis replied:
As an aside -- this all feels very hokey to me, I'd rather we just passed in the final set of substitutions, rather than passing in an expr-id and monomorphization substitution and then figuring out the substitutions. But that feels like a separate PR.

nikomatsakis commented:
so yes this looks overall good to me. I'd love to hear from @oli-obk, since he's been doing so much with constants, but I think I'm prepared to r+. I'm just trying to think if there is anything being unintentionally stabilized here -- I don't *think* so.

(I also agree that this is a distinct issue from what the RFC is addressing; this seems to be just pushing the current support for assoc constants further within the limits of the current design.)

oli-obk commented:
Since this only touches trans (const eval passes None for the param substs) it should be fine. But it might allow more than we want. Currently it should not allow referencing an assoc const from an assoc const. But if check_const doesn't catch it, trans might simply allow it. I'm not sure if there's a compile-fail test for it.

nikomatsakis commented:
@oli-obk my biggest worry are things being allowed without feature gates -- but it's good to identify bugs or deviations from the expected behavior. Everything I looked at seemed like it would be within bounds though. Why do you say it should not allow referencing an assoc. const from another assoc. const?

oli-obk commented:
Since the entire thing is behind a feature gate it is ok. But if we allow things in assoc constants that we don't allow in true const eval, it will be much harder getting them to work with array lengths in the future.

nikomatsakis commented:
@oli-obk 

> But if we allow things in assoc constants that we don't allow in true const eval, it will be much harder getting them to work with array lengths in the future.

I'm confused by this. I think you are saying: "early const eval (before trans) cannot handle assoc constants. If we want to allow associated constants to be handled by early const eval, we should ensure that they do not reference other associated constants, because those are not handled by early const eval". But it seems like they *would* be handled by early const eval if... they were handled by early const eval? 

But in any case I think that for assoc constants to be allowed in array lengths, we would have to move beyond the notion of early/late const eval, which is (imo) incorrect. Basically, we would sometimes have abstract array lengths that we can only determine precisely at trans time (but we may still know that two array lengths are the same, even if we don't know what they are yet).

oli-obk commented:
That's not the issue. The issue is things like addresses to statics which can't be handled. But you are right. My argument is circular.

To the original question. I cannot see a way this escapes a feature gate.

oli-obk commented on src/test/compile-fail/associated-const-type-parameter-arrays.rs:
```diff
@@ -11,16 +11,23 @@
-    <T as Foo>::MIN //~ ERROR E0329
+pub fn test<A: Foo, B: Foo>() {
+    let _array: [u32; <A as Foo>::Y] = //~ error: the parameter type
+        [4; <A as Foo>::Y];
```
Does the previous error prevent the repeat expr from emitting an error? If so, this probably needs a second test

nikomatsakis commented:
OK, I'm inclined to r+ for now I think, @michaelwu can you produce a rebased version that addresses the nits and answers various questions?

michaelwu commented:
Will do next week - on PTO right now.

michaelwu commented:
I've also rebased and squashed, and am building and testing now. These commits are probably easier to check for now.

michaelwu commented:
Pushed the rebase now since it looks like travis isn't happy.

nikomatsakis commented:
Ah, I forgot about that trait selection thing. Let me look at that a bit more closely.

nikomatsakis commented:
OK, so, I think that converting from `span_bug` to `None` looks correct. I don't see any tests of what happens when you have an invalid assoc const references, but it does seem to fail during typeck (e.g., with this test I wrote:

```rust
// Copyright 2015 The Rust Project Developers. See the COPYRIGHT
// file at the top-level directory of this distribution and at
// http://rust-lang.org/COPYRIGHT.
//
// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
// option. This file may not be copied, modified, or distributed
// except according to those terms.

#![feature(associated_consts)]

trait Foo {
    const ID: i32;
}

const X: i32 = <i32>::ID; //~ ERROR E0034

fn main() {
    assert_eq!(1, X);
}
```

it'd be great to make more tests of this ilk, I couldn't find any. e.g., using such constants in "early resolution" contexts, like a match arm.

nikomatsakis commented:
@michaelwu would you mind adding that test I pasted above into your PR, as well as the following one:

```rust
// Copyright 2015 The Rust Project Developers. See the COPYRIGHT
// file at the top-level directory of this distribution and at
// http://rust-lang.org/COPYRIGHT.
//
// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
// option. This file may not be copied, modified, or distributed
// except according to those terms.

#![feature(associated_consts)]

trait Foo {
    const ID: usize;
}

const X: [i32; <i32 as ID>::usize] = [0, 1, 2]; //~ ERROR E0250

fn main() {
    assert_eq!(1, X);
}
```

(I know it's sort of a pre-existing problem and not the fault of your PR, but it makes me feel better about that diff. ;)

michaelwu commented:
No problem - I've added the tests. I bumped the copyright year for both of them and modified the 2nd test a bit since it didn't look quite right and produced 4 errors. (`<i32 as ID>::usize` -> `<i32 as Foo>::ID`)

nikomatsakis commented:
@bors r+

bors commented:
:pushpin: Commit a4f91e5 has been approved by `nikomatsakis`

<!-- @bors r=nikomatsakis a4f91e5fed85164a8535a2300ca6445520d25059 -->

nikomatsakis commented:
Thanks @michaelwu (and @oli-obk for the feedback)

bors commented:
:hourglass: Testing commit a4f91e5 with merge 683af0d...

bors commented:
:sunny: Test successful - [auto-linux-32-nopt-t](http://buildbot.rust-lang.org/builders/auto-linux-32-nopt-t/builds/7703), [auto-linux-32-opt](http://buildbot.rust-lang.org/builders/auto-linux-32-opt/builds/7724), [auto-linux-64-debug-opt](http://buildbot.rust-lang.org/builders/auto-linux-64-debug-opt/builds/974), [auto-linux-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-linux-64-nopt-t/builds/7661), [auto-linux-64-opt](http://buildbot.rust-lang.org/builders/auto-linux-64-opt/builds/7678), [auto-linux-64-x-android-t](http://buildbot.rust-lang.org/builders/auto-linux-64-x-android-t/builds/7663), [auto-linux-cross-opt](http://buildbot.rust-lang.org/builders/auto-linux-cross-opt/builds/882), [auto-linux-musl-64-opt](http://buildbot.rust-lang.org/builders/auto-linux-musl-64-opt/builds/2833), [auto-mac-32-opt](http://buildbot.rust-lang.org/builders/auto-mac-32-opt/builds/7727), [auto-mac-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-mac-64-nopt-t/builds/7723), [auto-mac-64-opt](http://buildbot.rust-lang.org/builders/auto-mac-64-opt/builds/7687), [auto-win-gnu-32-nopt-t](http://buildbot.rust-lang.org/builders/auto-win-gnu-32-nopt-t/builds/2707), [auto-win-gnu-32-opt](http://buildbot.rust-lang.org/builders/auto-win-gnu-32-opt/builds/2722), [auto-win-gnu-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-win-gnu-64-nopt-t/builds/2721), [auto-win-gnu-64-opt](http://buildbot.rust-lang.org/builders/auto-win-gnu-64-opt/builds/2699), [auto-win-msvc-32-opt](http://buildbot.rust-lang.org/builders/auto-win-msvc-32-opt/builds/1941), [auto-win-msvc-64-opt](http://buildbot.rust-lang.org/builders/auto-win-msvc-64-opt/builds/2703)
